rust/lance-file/src/v2/reader.rs

```diff
@@ -0,0 +1,515 @@
+use std::{io::Cursor, ops::Range, sync::Arc};
+
+use arrow_array::RecordBatch;
+use arrow_schema::Schema;
+use byteorder::{ByteOrder, LittleEndian, ReadBytesExt};
+use bytes::{Bytes, BytesMut};
+use futures::{stream::BoxStream, StreamExt};
+use lance_encoding::{
+    decoder::{BatchDecodeStream, ColumnInfo, DecodeBatchScheduler, PageInfo},
+    EncodingsIo,
+};
+use prost::Message;
+use snafu::{location, Location};
+
+use lance_core::{Error, Result};
+use lance_encoding::format::pb as pbenc;
+use lance_io::{scheduler::FileScheduler, ReadBatchParams};
+use tokio::{sync::mpsc, task::JoinHandle};
+
+use crate::{
+    datatypes::{Fields, FieldsWithMeta},
+    format::{pb, pbfile, MAGIC, MAJOR_VERSION, MINOR_VERSION_NEXT},
+};
+
+use super::io::LanceEncodingsIo;
+
+// TODO: Caching
+pub struct CachedFileMetadata {
+    /// The schema of the file
+    pub file_schema: Schema,
+    /// The column metadatas
+    pub column_metadatas: Vec<pbfile::ColumnMetadata>,
+    /// The number of rows in the file
+    pub num_rows: u64,
+}
+
+pub struct FileReader {
+    scheduler: Arc<LanceEncodingsIo>,
+    file_schema: Schema,
+    column_infos: Vec<ColumnInfo>,
+    num_rows: u64,
+}
+
+struct Footer {
+    column_meta_start: u64,
+    // We don't use this today because we get here by subtracting backward from global_buf_start
+    #[allow(dead_code)]
+    column_meta_offsets_start: u64,
```

Comment:
Is there a reason not to use this? Is there any risk this becomes a bug later?

Reply:
I can update this comment.  We will need this if/when we want to support "true column projection" where we don't load any of the metadata for columns we aren't interested in.

I'm not sure we will want this anytime soon because then:
* If you are not using cached metadata then you are adding at least one extra IOP (first read footer, then read column metadata offsets table, then read column metadatas of interest, this last step is only one IOP of you coalesce which is non trivial)
* If you are using cached metadata then your metadata caching gets trickier (what happens if you have cached metadata for the file for columns X, and Z and the user now wants column Y?)  You will need to cache on a per-column basis and I wanted to avoid that complexity for now.